### PR TITLE
[RISCV] Support Global Dynamic TLSDESC in the RISC-V backend

### PR DESCRIPTION
**Pull request by ilovepi**
This patch adds basic TLSDESC support for the global dynamic case in the RISC-V backend.

Specifically, we add new relocation types for TLSDESC, as prescribed in 
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/pull/373, and add a new pseudo instruction 
to simplify code generation.

This patch does not try to optimize the local dynamic case, which can be improved in separate
patches. Linker side changes will also be handled separately.

The current implementation is only enabled when passing the new `-enable-tlsdesc` codegen flag.

### CODE REVIEW

**Comment by ilovepi** [comment]
ping

**Comment by ilovepi** [comment]
ping. @topperc @jrtc27 @asb can we take a look at this? psABI has approved TLSDESC, and there are patches out for GCC and binutils.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 1]
```diff
@@ -3107,6 +3136,41 @@ void RISCVAsmParser::emitLoadTLSGDAddress(MCInst &Inst, SMLoc IDLoc,
...
+                          RISCVMCExpr::VK_RISCV_TLSDESC_LOAD_LO, Ctx);
+
+  emitToStreamer(Out, MCInstBuilder(RISCV::ADDI)
+                          .addOperand(DestReg)
+                          .addOperand(DestReg)
+                          .addExpr(RefToLinkTmpLabel));
+
+  emitToStreamer(
+      Out,
+      MCInstBuilder(RISCV::JALR).addReg(RISCV::X5).addReg(RISCV::X5).addImm(0));
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3171]: Should this use DestReg instead of X5?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 2]
```diff
@@ -3107,6 +3136,41 @@ void RISCVAsmParser::emitLoadTLSGDAddress(MCInst &Inst, SMLoc IDLoc,
...
 }
 
+void RISCVAsmParser::emitLoadTLSDescAddress(MCInst &Inst, SMLoc IDLoc,
+                                            MCStreamer &Out) {
+  // The load TLS GD address pseudo-instruction "la.tlsdesc" is used in
+  // global-dynamic TLS model addressing of global symbols:
+  //   la.tlsdesc rdest, symbol
+  // expands to
+  //   TmpLabel: AUIPC rdest, %tlsdesc_hi(symbol)
+  //             ADDI rdest, rdest, %pcrel_lo(TmpLabel)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3146]: what about the JALR?

**Review by topperc — COMMENTED** [review]
Do we need a test in test/MC/RISCV for the AsmParser?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 3]
```diff
@@ -3246,6 +3310,18 @@ bool RISCVAsmParser::checkPseudoAddTPRel(MCInst &Inst,
 
   return false;
 }
+bool RISCVAsmParser::checkPseudoTLSDESCCall(MCInst &Inst,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3313]: Does this have a caller?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 2 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3146]: Actually now that you’ve mentioned that bug in the documentation I’m questioning whether using the pseudo instruction is the correct approach. I followed the example for the existing TLS implementations, but those are also mentioned in the psABI and I think this may have been my own invention when I was first learning my way around the backend (and still am). To work correctly I think the pseudo would either need another register or I guess assemblers would need to scavenge one.  Either way I don’t think that is quite right. Well … unless there is a way to make a pseudo (or something like) it completely internal, which I’m not aware of. 

In that case, I guess I would need to translate the pseudo expansion into SelectionDag code instead.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3146]: So, AArch64 has the pseudo `TLSDESC_CALLSEQ`, and nothing in the AsmParser. They do have a translation in the AsmPrinter though, which I don't think is functionally different than the pseudo expansion. 

Do yo think its appropriate to omit the AsmParser bits here, since I don't expect for us to ever emit `la.tlsdesc` instructions? Otherwise, I'm not super sure how to adapt this, since its unclear to me how to write the code sequence, labels, and relocations in RISCVISelLowering.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 3 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3313]: I believe I needed to implement this for something in that tablegen created. Let me double check that though.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3313]: Ah, that wasn't it.  I am missing the call in ProcessInstruction,  similar to the AddTPRel case.  Thanks for pointing this out.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def [hunk 4]
```diff
@@ -55,3 +55,7 @@ ELF_RELOC(R_RISCV_SET32,             56)
 ELF_RELOC(R_RISCV_32_PCREL,          57)
 ELF_RELOC(R_RISCV_IRELATIVE,         58)
 ELF_RELOC(R_RISCV_PLT32,             59)
+ELF_RELOC(R_RISCV_TLSDESC_HI20,      62)
```
Comment (topperc) [inline llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def:58]: I think this needs to be rebased. There should be a 60 and 61 defined.

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp [hunk 5]
```diff
@@ -539,6 +545,53 @@ bool RISCVPreRAExpandPseudo::expandLoadTLSGDAddress(
...
+      .addReg(DestReg)
+      .addImm(0)
+      .addSym(AUIPCSymbol, RISCVII::MO_TLSDESC_CALL);
+
+  MachineInstr *TPAdd = BuildMI(MBB, MBBI, DL, TII->get(RISCV::ADD), FinalReg)
+                            .addReg(RISCV::X10)
+                            .addReg(RISCV::X4);
+
+  if (MI.hasOneMemOperand())
+    TPAdd->addMemOperand(*MF, *MI.memoperands_begin());
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:590]: I don't think an ADD instruction should have a memory operand.

I'm not sure where a MemOperand on the MI would come from, there wasn't one created in getGeneralDynamicTLSDescAddr.

**Comment by topperc** [comment]
Do these instructions need to appear back to back in the final binary. I notice AArch mentions a linker optimization. Expanding in pre RA expansion would allow them to become separated.

File: llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp (hunk 5 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:590]: That's a good point.  I'll update the patch to reflect that.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def (hunk 4 above)
Comment (ilovepi) [reply llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def:58]: Yes, it does. Did we resolve how to do that w/ PRs? I recall there was some concerns about the rebase causing issues with code review.

My understanding is that `git rebase main` and pushing to my branch will erase much of the comments and review history. Do you know if there is a good way to rebase w/o losing the review comments? I've been following https://discourse.llvm.org/t/force-push-and-rebase/73766 but I don't see a concrete reccomendation in the thread.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 2 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp]: I've gone ahead and dropped the AsmParser bits for `la.tlsdesc`, since we won't output that into assembly. Le tme now if you're happy with that change, or you'd prefer a different solution.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 3 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:3345]: done

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 6]
```diff
@@ -6843,6 +6845,24 @@ SDValue RISCVTargetLowering::getDynamicTLSAddr(GlobalAddressSDNode *N,
   return LowerCallTo(CLI).first;
 }
 
+SDValue
+RISCVTargetLowering::getGeneralDynamicTLSDescAddr(GlobalAddressSDNode *N,
```
Comment (MaskRay) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:6849]: `GeneralDynamicTLSDesc` in the name` `getGeneralDynamicTLSDescAddr` is misleading. general-dynamic refers to a traditional TLS mode while TLSDESC is entirely different.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:6849]: Well, AArch64 handles lowering for llvm's general-dynamic and local-dynamic cases differently. I was trying to make it clear here that we're not trying to do anything fancy like elide loads of the `__MODULE_BASE__` for multiple accesses.  I'm not convinced we want to do that in RISC-V, so it's probably fine to rename this.

Maybe `getTLSDescAddr` is good enough? at least until we decide that we want to optimize them differently.

Comment (MaskRay) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:6849]: getTLSDescAddr sounds good.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: done.

File: llvm/test/MC/RISCV/rv32i-invalid.s [hunk 7]
```diff
@@ -188,3 +188,8 @@ addi a2, ft0, 24 # CHECK: :[[@LINE]]:10: error: invalid operand for instruction
 
 # fence.tso accepts no operands
 fence.tso rw, rw # CHECK: :[[@LINE]]:11: error: invalid operand for instruction
+
+.Ltlsdesc_hi0:
+jalr	x5, 0(a1), %tlsdesc_hi(.Ltlsdesc_hi0) # CHECK: :[[@LINE]]:17: error: operand must be a symbol with %tlsdesc_call modifier
+jalr	x1, 0(a1), %tlsdesc_call(.Ltlsdesc_hi0) # CHECK: :[[@LINE]]:12: error: the output operand must be t0/x5 when using %tlsdesc_call modifier
```
Comment (topperc) [inline llvm/test/MC/RISCV/rv32i-invalid.s:194]: Looks like we are missing MC layer tests that don't generate an error?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 8]
```diff
@@ -1722,6 +1722,35 @@ let hasSideEffects = 0, mayLoad = 0, mayStore = 0, Size = 8, isCodeGenOnly = 0,
     isAsmParserOnly = 1 in
 def PseudoLA_TLS_GD : Pseudo<(outs GPR:$dst), (ins bare_symbol:$src), [],
                              "la.tls.gd", "$dst, $src">;
+let hasSideEffects = 0, mayLoad = 1, mayStore = 0, Size = 32, isCodeGenOnly = 0,
+    isAsmParserOnly = 1 in
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:1726]: This says AsmParserOnly, but I don't see any assembler tests for it.

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 9]
```diff
@@ -597,7 +613,10 @@ struct RISCVOperand final : public MCParsedAsmOperand {
     if (!isImm())
       return false;
     bool IsConstantImm = evaluateConstantImm(getImm(), Imm, VK);
-    if (VK == RISCVMCExpr::VK_RISCV_LO || VK == RISCVMCExpr::VK_RISCV_PCREL_LO)
+    if (VK == RISCVMCExpr::VK_RISCV_LO ||
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:616]: None of these change to accept more values for `VK` are tested.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h [hunk 10]
```diff
@@ -71,6 +71,18 @@ enum Fixups {
   // Used to generate an R_RISCV_ALIGN relocation, which indicates the linker
   // should fixup the alignment after linker relaxation.
   fixup_riscv_align,
+  // 20-bit fixup corresponding to %tlsdesc_hi(foo) for instructions like
+  // auipc
+  fixup_riscv_tlsdesc_hi20,
+  // 12-bit fixup corresponding to %tlsdesc_load_lo(foo) for instructions like
+  // addi
```
Comment (topperc) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h:78]: Should this say load instead of addi?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 8 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td]: Ah, I think I missed this when I dropped the AsmParser bits.  So, I *think* that means I want to set isAsmParserOnly=0, isCodeGenOnly=0, isPseudo=1, which I think is the case now. Is that the correct way to model a pseudo that is only for some internal state that shouldn't be emitted to assembly? 

That's my understanding from https://github.com/llvm-mirror/llvm/blob/2c4ca6832fa6b306ee6a7010bfb80a3f2596f824/include/llvm/Target/Target.td#L554 , at least, but I'm not sure if that is a correct interpretation or not.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h (hunk 10 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h]: Yes, I've fixed it now to use `lw`, but maybe it should just read `... for load instructions` instead? WDYT?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp (hunk 9 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:616]: I think the tests added in RISCV/MC should cover this now. Please let me know if I've missed one of the cases.

File: llvm/test/MC/RISCV/rv32i-invalid.s (hunk 7 above)
Comment (ilovepi) [reply llvm/test/MC/RISCV/rv32i-invalid.s:194]: You're right. I've added some that should address that i `rv32-tlsdesc-valid.s` and `rv64-tlsdesc-valid.s`

File: llvm/test/MC/RISCV/rv32zfh-invalid.s [hunk 11]
```diff
@@ -3,8 +3,8 @@
 
 # Out of range immediates
 ## simm12
-flh ft1, -2049(a0) # CHECK: :[[@LINE]]:10: error: operand must be a symbol with %lo/%pcrel_lo/%tprel_lo modifier or an integer in the range [-2048, 2047]
-fsh ft2, 2048(a1) # CHECK: :[[@LINE]]:10: error: operand must be a symbol with %lo/%pcrel_lo/%tprel_lo modifier or an integer in the range [-2048, 2047]
+flh ft1, -2049(a0) # CHECK: :[[@LINE]]:10: error: operand must be a symbol with %lo/%pcrel_lo/%tprel_lo/%tlsdesc_load_lo modifier or an integer in the range [-2048, 2047]
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/rv32zfh-invalid.s:6]: `%tlsdesc_load_lo` only makes sense for `ld`. Changing the diagnostic will require updates to lots of files and is misleading to a lot of instructions. Perhaps the initial patch can drop this message update.

File: llvm/test/MC/RISCV/tlsdesc.s [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	# INST: addi a0, a0, 0x0
+	jalr	t0, 0(a1), %tlsdesc_call(.Ltlsdesc_hi0)
+	# INST: jalr t0, 0x0(a1)
+	add	a0, a0, tp
+	# INST: add a0, a0, tp
+	ret
+
+## Check invalid usage
+.ifdef ERR
+	auipc x1, %tlsdesc_call(foo) # ERR: :[[@LINE]]:12: error: operand must be a symbol with a %pcrel_hi/%got_pcrel_hi/%tls_ie_pcrel_hi/%tls_gd_pcrel_hi/%tlsdesc_hi modifier or an integer in the range
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/tlsdesc.s:31]: Note: `[[@LINE]]` is deprecated. Use `[[#@LINE]]`

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h [hunk 13]
```diff
@@ -71,6 +71,18 @@ enum Fixups {
   // Used to generate an R_RISCV_ALIGN relocation, which indicates the linker
   // should fixup the alignment after linker relaxation.
   fixup_riscv_align,
+  // 20-bit fixup corresponding to %tlsdesc_hi(foo) for instructions like
+  // auipc
+  fixup_riscv_tlsdesc_hi20,
```
Comment (MaskRay) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h:76]: > instructions like auipc

It seems that tlsdesc is auipc-style only, instead of lui-style. You can say that this is for auipc.

File: llvm/test/MC/RISCV/rv32i-invalid.s (hunk 7 above)
Comment (MaskRay) [reply llvm/test/MC/RISCV/rv32i-invalid.s:194]: I think it's clearer to place these negative tests in a test named `tlsdesc*.s`

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h (hunk 13 above)
Comment (MaskRay) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h:76]: I wonder whether we can just use `MCFixupKind(FirstLiteralRelocationKind + ELF::R_RISCV_TLSDESC_CALL)`

File: llvm/test/MC/RISCV/rv32zfh-invalid.s (hunk 11 above)
Comment (ilovepi) [reply llvm/test/MC/RISCV/rv32zfh-invalid.s]: done.

File: llvm/test/MC/RISCV/tlsdesc.s (hunk 12 above)
Comment (ilovepi) [reply llvm/test/MC/RISCV/tlsdesc.s]: Thanks. done.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h (hunk 13 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVFixupKinds.h:76]: I've adjusted the comments per your suggestion, but could you elaborate more on what you're thinking w.r.t. `MCFixupKind(FirstLiteralRelocationKind + ELF::R_RISCV_TLSDESC_CALL)`?

**Comment by ilovepi** [comment]
@topperc @MaskRay , do either of you have strong feelings about abandoning this in favor of https://github.com/llvm/llvm-project/pull/77515? That PR was made w/ `spr` and should be a bit easier to understand/review (since it won't have merge or fixup commits) and will make the LLD side changes easier to stack.

**Comment by ilovepi** [comment]
As @MaskRay  pointed out in #66916, there's a bit too much context in this review, so please just ignore my earlier question about moving to a stacked PR.

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp [hunk 14]
```diff
@@ -80,6 +80,11 @@ static cl::opt<bool> EnableRISCVDeadRegisterElimination(
              " them with stores to x0"),
     cl::init(true));
 
+// TODO: This should be controlled by -mtls-dialect=<option>
+cl::opt<bool> EnableRISCVTLSDESC("riscv-enable-tlsdesc",
```
Comment (MaskRay) [inline llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:84]: We should add an option to `codegen::InitTargetOptionsFromCodeGenFlags`, then reference it using something like `DAG.getTarget().useTLSDESC()`. aarch64 ELF only implements TLSDESC for dynamic TLS models, it is fine for it to ignore `useTLSDESC`.

**Review by MaskRay — APPROVED** [review]
LGTM once the TLSDESC option is moved to use llvm/lib/CodeGen/CommandFlags.cpp

It's worth giving others some time to respond.

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp (hunk 14 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVTargetMachine.cpp]: That's a good suggestion. I was thinking of doing something like that as a follow up for the Clang driver patches, but I agree, its probably better to do it in this patch.

**Comment by ilovepi** [comment]
@topperc Is there anything else you think needs to be done for the code generation?

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp [hunk 15]
```diff
@@ -154,6 +158,32 @@ void RISCVMCCodeEmitter::expandFunctionCall(const MCInst &MI,
...
+  MCRegister Link = MI.getOperand(0).getReg();
+  MCRegister Dest = MI.getOperand(1).getReg();
+  MCRegister Imm = MI.getOperand(2).getImm();
+  Fixups.push_back(MCFixup::create(
+      0, Expr, MCFixupKind(RISCV::fixup_riscv_tlsdesc_call), MI.getLoc()));
+  // Emit fixup_riscv_relax for jalr where the relax feature is enabled.
+  if (STI.hasFeature(RISCV::FeatureRelax)) {
+    const MCConstantExpr *Dummy = MCConstantExpr::create(0, Ctx);
+    Fixups.push_back(MCFixup::create(
+        0, Dummy, MCFixupKind(RISCV::fixup_riscv_relax), MI.getLoc()));
```
Comment (MaskRay) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp:178]: I think we should remove this `fixup_riscv_relax`.

The spec says that the 4 instructions all have `R_RISCV_RELAX` in the presence of relaxation. The patch currently adds just one `R_RISCV_RELAX`, which is strange. I think it's fair to not support linker relaxation, which is probably not useful anyway.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 16]
```diff
@@ -89,8 +89,8 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
   if ((ABI == RISCVABI::ABI_ILP32F || ABI == RISCVABI::ABI_LP64F) &&
       !Subtarget.hasStdExtF()) {
     errs() << "Hard-float 'f' ABI can't be used for a target that "
-                "doesn't support the F instruction set extension (ignoring "
-                          "target-abi)\n";
+              "doesn't support the F instruction set extension (ignoring "
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:92]: Please don't reformat unrelated code.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:92]: ah, sorry. I had touched these lines in an earlier patch and hadn't realized I kept the formatting change. I'll fix this shortly.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: There is a lot of re-formatting to that file, not just the line I commented on.

Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: oh, I see it now.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def [hunk 17]
```diff
@@ -56,3 +56,7 @@ ELF_RELOC(R_RISCV_IRELATIVE,         58)
 ELF_RELOC(R_RISCV_PLT32,             59)
 ELF_RELOC(R_RISCV_SET_ULEB128,       60)
```
Comment (MaskRay) [inline llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def:57]: Define the dynamic relocation type R_RISCV_TLSDESC  12 as well.

(The lld patch incorrectly uses `R_RISCV_TLSDESC_CALL` for the dynamic reloc. I am going to take over that part and TLS optimization.)

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 16 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Sorry for that mess. I normally only use `git clang-format HEAD~`, but I must have use some other command or the `clangd` integration by mistake. I think this should be fixed. I'm looking at other files, and haven't seen anything obvious yet.

Comment (MaskRay) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Ah, I only use `git diff -U0 --no-color --relative main... -- | path/to/clang/tools/clang-format/clang-format-diff.py -p1  -i` :)

**Comment by topperc** [comment]
code generation looks good to me, but I guess you have some comments from MaskRay to address now.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def (hunk 17 above)
Comment (ilovepi) [reply llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def:58]: I've removed them per your request. Just so we'd have a version where they were in place, I added the code to make them relaxable. But I'm fine either way, so if you'd prefer we start w/o the relaxation relocations it should be good now.

Related: do you think its worth testing that they're *not* there?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 16 above)
Comment (ilovepi) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Thanks. I'll try out that incantation tomorrow. :)

**Comment by MaskRay** [comment]
This PR needs a rebase.. There are quite a few merge commits. Hmm, I don't know how to squash the changes to rebase them to latest main branch. `git rebase --keep-base -i main` has quite a few merge conflicts.

(My lld change (https://github.com/maskray/llvm-project/tree/rv-tlsdesc) is almost complete. I want to rebase it onto the latest revision of this LLVM PR.)

**Comment by ilovepi** [comment]
@MaskRay I've rebased it for you.

**Comment by MaskRay** [comment]
> The current implementation is only enabled when passing the new `-riscv-enable-tlsdesc` flag.

This needs an update. The cl::opt has been removed.

File: llvm/test/MC/RISCV/tlsdesc.s [hunk 18]
```diff
@@ -0,0 +1,44 @@
+# RUN: llvm-mc -filetype=obj -triple riscv32 < %s --defsym RV32=1  | llvm-objdump -d -M no-aliases - | FileCheck %s --check-prefixes=INST,RV32
+# RUN: llvm-mc -filetype=obj -triple riscv64 < %s | llvm-objdump -d -M no-aliases - | FileCheck %s --check-prefixes=INST,RV64
```
Comment (MaskRay) [inline llvm/test/MC/RISCV/tlsdesc.s:2]: I think we should use llvm-objdump `-d -r` to check relocations.

File: llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def (hunk 17 above)
Comment (MaskRay) [reply llvm/include/llvm/BinaryFormat/ELFRelocs/RISCV.def:58]: > Related: do you think its worth testing that they're not there?

I've made a comment on the test "I think we should use llvm-objdump -d -r to check relocations."
If you do that, `-NEXT` is sufficient to ensure that there is no `R_RISCV_RELAX` :)

With my experience on https://github.com/llvm/llvm-project/pull/79099 , I think two `R_RISCV_RELAX` are sufficient. I probably need to follow up on the spec.

File: llvm/test/MC/RISCV/tlsdesc.s (hunk 18 above)
Comment (ilovepi) [reply llvm/test/MC/RISCV/tlsdesc.s]: done

**Comment by ilovepi** [comment]
@topperc I spoke to @MaskRay and he's satified w/ the patch now. Is there anything else you'd like to see addressed? or are we good to land this?

**Review by topperc — APPROVED** [review]
LGTM
